File: algorithms/sorting/bubble-sort/bubble-sort.cpp
#include <iostream>

void
bubbleSort(int a[], int l)
{
    for (int x = 0; x < l - 1; ++x) {
    int s = true;
    for (int y = 0; y < l - x - 1; ++y) {
        int current = a[y];
        int next = a[y + 1];

        if (current > next) {
          a[y + 1] = current;
          a[y] = next;
          s = false;
        }
    }
    if (s)
      break;
    }
}

void
printArray(int a[], int l)
{
  int x = 0;
  while (x < l) {
    std::cout << a[x];
    std::cout << ", ";
    ++x;
  }
}


int main()
{
  //int arr[] = { 6, 4, 8, 3, 2, 9, 0, 1, 7, 6, 5 };
  int arr[] = { 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0 };
  int arrLength = sizeof(arr) / sizeof(arr[0]);

  std::cout << "Input Array:" << std::endl;
  printArray(arr, arrLength);
  std::cout << std::endl << std::endl;
  std::cout << "Output Array:" << std::endl;
  bubbleSort(arr, arrLength);
  printArray(arr, arrLength);

  return 0;
}

File: algorithms/sorting/bubble-sort/bubble-sort.php
<?php

function getTestArray($len)
{
    $tmpArr = range(0, $len - 1);
    shuffle($tmpArr);
    return $tmpArr;
}

function shouldSwap($a, $b, $dir)
{
    return $dir === "ASC" ? $a > $b : $a < $b;
}

function swap($keyA, $keyB, $arr)
{
    $tmpArr = array_replace([], $arr);
    [$tmpArr[$keyA], $tmpArr[$keyB]] = [$tmpArr[$keyB], $tmpArr[$keyA]];

    return $tmpArr;
}

function bubbleSort($arr, $dir)
{
    $cpyArr = array_replace([], $arr);
    $arrLength = count($cpyArr);

    for ($x = 0; $x < $arrLength; $x++)
        for ($y = 0; $y < ($arrLength - 1 - $x); $y++) {
            $currentVal = $cpyArr[$y];
            $nextVal = $cpyArr[$y + 1];
            if (shouldSwap($currentVal, $nextVal, $dir)) {
                $cpyArr = swap($y, $y + 1, $cpyArr);
            }
        }

    return $cpyArr;
}

$testArray = getTestArray(20);

echo "Input:\n";
print_r($testArray);

echo "\n";

echo "Output:\n";
print_r(bubbleSort($testArray, "ASC"));
